src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferContainsKeyOrContainsValue.cs

```diff
@@ -0,0 +1,108 @@
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var idictionaryType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIDictionary2);
+                var ireadOnlyDictionaryType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIReadOnlyDictionary2);
+
+                if (idictionaryType == null && ireadOnlyDictionaryType == null)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(context =>
+                {
+                    var invocation = (IInvocationOperation)context.Operation;
+
+                    if (invocation.TargetMethod.Name != "Contains" ||
+                        !(GetReceiverOperationOrDefault(invocation) is IPropertyReferenceOperation propertyReference) ||
+                        !(propertyReference.Member is IPropertySymbol property))
+                    {
+                        return;
+                    }
+
+                    if (property.Name == "Keys")
+                    {
+                        // The 'ContainsKey' method is part of the interface so we do not need extra checks
+                        if (IsDictionary(property, idictionaryType, ireadOnlyDictionaryType))
+                        {
+                            context.ReportDiagnostic(invocation.CreateDiagnostic(Rule));
+                        }
+                    }
+                    else if (property.Name == "Values")
+                    {
+                        // The 'ContainsValue' method only exists in some of the implementations so we need to check
+                        // if the method exists before reporting a diagnostic.
+                        if (IsDictionary(property, idictionaryType, ireadOnlyDictionaryType) &&
+                            property.ContainingType.GetMembers("ContainsValue").OfType<IMethodSymbol>().Any())
+                        {
+                            context.ReportDiagnostic(invocation.CreateDiagnostic(Rule));
+                        }
+                    }
+                    else
+                    {
+                        // do nothing
+                    }
+                }, OperationKind.Invocation);
+            });
+        }
+
+        private static bool IsDictionary(IPropertySymbol property, INamedTypeSymbol? idictionaryType, INamedTypeSymbol? ireadOnlyDictionaryType)
+            => property.ContainingType.OriginalDefinition.Equals(idictionaryType) ||
```

Comment:
I am wondering if it might be worth having a `IsOrDerivesFrom` extension method?